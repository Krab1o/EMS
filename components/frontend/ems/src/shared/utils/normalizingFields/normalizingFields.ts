import { isValid, parseISO } from 'date-fns';
import type { AnyObject } from 'antd/es/_util/type';

export function deNormalize(values: AnyObject | null) {
  const init: AnyObject = {};
  if (values) {
    Object.entries(values).forEach(([key, value]) => {
      const timeValue = parseISO(value);
      if (value && isValid(new Date(value)) && isValid(timeValue)) {
        init[key] = new Date(parseISO(value));
      } else {
        init[key] = value;
      }
    });
  }
  return init;
}

export function normalize<T>(values: T) {
  const finishValues: AnyObject = {};
  Object.entries(values as AnyObject).forEach(([key, value]) => {
    if (typeof value?.getMonth === 'function') {
      finishValues[key] = value ? value.toISOString() : undefined;
    } else {
      finishValues[key] = value;
    }
  });
  return finishValues;
}
