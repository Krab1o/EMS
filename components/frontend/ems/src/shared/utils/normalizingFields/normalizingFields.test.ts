import { normalize, deNormalize } from './normalizingFields';

const mockWithDate = {
  value1: '23',
  value2: 23,
  value3: undefined,
  date: new Date(2013, 3, 3, 3, 34, 14),
};

const mockWithDateString = {
  value1: '23',
  value2: 23,
  value3: undefined,
  date: '2013-04-02T23:34:14.000Z',
};

describe('normalizing', () => {
  test('Валидация объекта с датой', () => {
    expect(normalize(mockWithDate)).toEqual(mockWithDateString);
    expect(normalize(mockWithDate)).toMatchSnapshot();
  });
  test('Валидация объекта со строкой', () => {
    expect(deNormalize(mockWithDateString)).toEqual(mockWithDate);
    expect(deNormalize(mockWithDateString)).toMatchSnapshot();
  });
});
