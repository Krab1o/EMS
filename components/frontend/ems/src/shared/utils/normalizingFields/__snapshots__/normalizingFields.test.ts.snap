// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizing Валидация объекта с датой 1`] = `
Object {
  "date": "2013-04-02T23:34:14.000Z",
  "value1": "23",
  "value2": 23,
  "value3": undefined,
}
`;

exports[`normalizing Валидация объекта со строкой 1`] = `
Object {
  "date": 2013-04-02T23:34:14.000Z,
  "value1": "23",
  "value2": 23,
  "value3": undefined,
}
`;
